### DISCUSSION TITLE
Speech?

### DISCUSSION THREAD

**Discussion by tensiondriven** [gtxaspec/wz_mini_hacks#666] [Q&A]
Does wz_mini_hacks support text-to-speech?  I didn't see an entry under `cmd`.

When wz_mini_hacks first boots up, it speaks a little message.  I swear that about a year ago, this functionality was also exposed on installations, such that one could, for example, make the camera talk from a script.

This would be useful for things like Home Assistant automations, wherein one could get voice notifications based on events in HA.

Does that functionality still exist?  If not, how is the audio installation message generated?

**Comment by gtxaspec** [comment]
prerecorded files... not enough horsepower to generate voice, you could use something like https://github.com/MycroftAI/mimic3 with HA to do it, pretty easy and it works.

**Reply by endertable** [reply]
Hi @gtxaspec 

In the Documentation -> Notes.md, you have the following:

```
take a RAW snapshot (v2 only)

impdbg --save_pic /tmp/output.nv12 --pic_type NV12

convert to jpg:

ffmpeg -loglevel quiet -y -f rawvideo -pixel_format nv12 -s 1920x1080 -i /tmp/output.nv12 -vf fps=1 /media/mmc/output.jpg
```

The impdbg command also works on the PanV2, but the ffmpeg does not convert it to a jpg.  It errors out with:  
   
  Unknown input format: 'rawvideo'

Am I missing something or it this only for the ffmpeg compiled with atomcam?
Thanks

**Comment by endertable** [comment]
Is is just me or does finding just the right .wav file for the V3 to play correctly a PITA?  Is it like pcm_s16le or something, you almost have to reformat a .wav file to a specific format.

**Reply by gtxaspec** [reply]
Just use raw pcm, ffmpeg generates them fine.  `-f s16le -ar 16000 -ac 1`

**Comment by endertable** [comment]
@tensiondriven if you are looking to create a few predefined files, then you can try the following script to create wav files (woman's voice).  Simply place into a file (e.g. getT2S.sh) and run with the text you want.  It will save into /media/mmc/media as default.  It is meant to be run directly from a V3 or PanV2 cam.  Have not tested on V2 nor PanV1.  I wouldn't do excessive spammy runs or you may get blocked from the site, but can get you started, if this is kind of what you are looking for?

./getT2S.sh step away from the door


```
#!/bin/sh

SAVETO="/media/mmc/media"

if [ "$1" = "" ]
then
  echo Must supply some text
  exit 0
fi

if ! [ -d $SAVETO ]
then
  mkdir $SAVETO
  if [ $? != 0 ]
  then
    echo ERROR creating $SAVETO !  Maybe SD Card not mounted
    exit 0
  fi
fi

TEXTPOST=$(echo $@ | sed "s/ /+/g")
TEXTNAME=$(echo $@ | sed "s/ /_/g")

rm -f /run/TTS.runner /run/TTS_tmp.mp3

curl 'https://www.texttospeechfree.com/' \
  -H 'authority: www.texttospeechfree.com' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'cache-control: max-age=0' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'origin: https://www.texttospeechfree.com' \
  -H 'referer: https://www.texttospeechfree.com/' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36 Edg/118.0.2088.76' \
  --data "word=${TEXTPOST}&language1=en-us&audio_format=24khz_8bit_stereo&audiocateg=%25s.mp3&submit=Play" > /run/TTS.runner

URL=$(grep -o 'https://www.texttospeechfree.com/TextToSpeech[^"]*' /run/TTS.runner)
echo URL is $URL

if [ "$URL" != "" ]
then
  curl "${URL}" \
  -H 'authority: www.texttospeechfree.com' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'cache-control: max-age=0' \
  -H 'origin: https://www.texttospeechfree.com' \
  -H 'referer: https://www.texttospeechfree.com/' \
  -H 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36 Edg/118.0.2088.76' > /run/TTS_tmp.mp3
else
  echo Did not catch a URL
  exit 0
fi 
  
ffmpeg -i /run/TTS_tmp.mp3 -f s16le -ar 16000 -ac 1 $SAVETO/${TEXTNAME}.wav
audioplay_t31 $SAVETO/${TEXTNAME}.wav 80
rm -f /run/TTS.runner /run/TTS_tmp.mp3
```
